lib/dslash_quda.cu

```diff
@@ -1,63 +1,27 @@
-#include <stack>
-
-#include <color_spinor_field.h>
-#include <clover_field.h>
-#include <dslash_quda.h>
-#include <color_spinor_field_order.h>
-#include <clover_field_order.h>
-#include <index_helper.cuh>
-#include <color_spinor.h>
-#include <linalg.cuh>
-#include <dslash_policy.cuh>
-#include <instantiate.h>
 #ifdef NVSHMEM_COMMS
 #include <cuda/atomic>
 #endif
 
-namespace quda {
-
-  // these should not be namespaced!!
-  // determines whether the temporal ghost zones are packed with a gather kernel,
-  // as opposed to multiple memcpys
-  static bool kernelPackT = false;
-
-  void setKernelPackT(bool packT) { kernelPackT = packT; }
-
-  bool getKernelPackT() { return kernelPackT; }
-
-  static std::stack<bool> kptstack;
-
-  void pushKernelPackT(bool packT)
-  {
-    kptstack.push(getKernelPackT());
-    setKernelPackT(packT);
-
-    if (kptstack.size() > 10)
-    {
-      warningQuda("KernelPackT stack contains %u elements.  Is there a missing popKernelPackT() somewhere?",
-      static_cast<unsigned int>(kptstack.size()));
-    }
-  }
+#include "color_spinor_field.h"
+#include "dslash_quda.h"
+#include <dslash_policy.cuh>
+#include "tunable_nd.h"
+#include "instantiate.h"
+#include "kernels/dslash_shmem_helper.cuh"
 
-  void popKernelPackT()
-  {
-    if (kptstack.empty())
-    {
-      errorQuda("popKernelPackT() called with empty stack");
-    }
-    setKernelPackT(kptstack.top());
-    kptstack.pop();
-  }
+namespace quda {
 
   namespace dslash {
     int it = 0;
 
-    cudaEvent_t packEnd[2];
-    cudaEvent_t gatherStart[Nstream];
-    cudaEvent_t gatherEnd[Nstream];
-    cudaEvent_t scatterStart[Nstream];
-    cudaEvent_t scatterEnd[Nstream];
-    cudaEvent_t dslashStart[2];
+    static constexpr int nDim = 4;
+    static constexpr int nDir = 2;
+    static constexpr int nStream = nDim * nDir + 1;
```

Comment:
This is also defined in `lib/targets/cuda/device.cpp` (for example) --- is there a reason why the definition here and the definition in `device.cpp` can't be unified as some "global" `constexpr` function defined in `include/device.h` or something of the sort?

Reply:
Ok, as noted below, these are two distinct `nStream` parameters, even if they have the same value.